Support Iceberg's DROP TABLE for corrupted tables

krvikash opened the pull request:
<!-- Thank you for submitting a pull request! Find more information in our development guide at https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md and contact us on #dev in Slack. -->
<!-- Provide an overview of the PR for maintainers and reviewers. -->
## Description

Fixes https://github.com/trinodb/trino/issues/12318

Follow up of https://github.com/trinodb/trino/pull/16651 for corrupted Iceberg Tables

<!-- Mark the appropriate option with an (x). Propose a release note if you can. -->
## Release notes

(X) Release notes are required, with the following suggested text:

```markdown
# Iceberg
* Support Iceberg's DROP TABLE for corrupted tables. ({issue}`16674`)
```

findinpath commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/AbstractIcebergTableOperations.java:
```diff
@@ -226,13 +229,25 @@ protected void refreshFromMetadataLocation(String newLocation)
+        catch (RuntimeException e) {
+            if (e.getCause() != null) {
+                if (e.getCause() instanceof FileNotFoundException
+                        || e.getCause().getMessage().contains("The specified key does not exist")) {
```
The check against an error message is not ideal.
For S3, look into `TrinoS3FileSystem` whether it would be appropriate to throw a `FileNotFoundException` in such cases 

`io.trino.plugin.hive.s3.TrinoS3FileSystem.TrinoS3InputStream#read(long, byte[], int, int)`

findepi replied:
what was the exception previously raised here?

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:
```diff
@@ -2691,20 +2722,25 @@ else if (strings.size() != 2) {
                 return new MaterializedViewFreshness(STALE, Optional.empty());
             }
+
+            if (tableHandle instanceof CorruptedIcebergTableHandle) {
```
combine with `if (tableHandle == null) {` preceding case (updating the comment message inside)

krvikash commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestCorruptedIcebergTable.java:
```diff
@@ -0,0 +1,134 @@
+
+        // Assert queries fail cleanly
+        assertThatThrownBy(() -> query("TABLE " + tableName))
+                .hasMessageContaining("Failed to open input stream for file:");
```
Yes, It was intentional in 1st commit `Test Iceberg connector behavior for a corrupted table`. The exception thrown is not a `TrinoException`, that is why I can not use `assertQueryFailure` and end up using `assertThatThrownBy`.

However, in the 2nd commit, I replaced `assertThatThrownBy` with `assertQueryFailure`.

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalog.java:
```diff
@@ -356,16 +357,54 @@ public Table loadTable(ConnectorSession session, SchemaTableName table)
     {
-        BaseTable table = (BaseTable) loadTable(session, schemaTableName);
-        validateTableCanBeDropped(table);
+        com.amazonaws.services.glue.model.Table metastoreTable;
```
Why do we need any changes to the Drop table logic in a catalog implementation?

perhaps, dropping corrupted table could be done separately, as an unregister_table + delete table directory.
if the table is corrupted, we probably should not expect catalog do anything smarter than that

cc @alexjo2144 @findinpath @electrum thoughts?

krvikash replied:
> unregister_table + delete table directory.

I like this approach. I was concerned when `data files of the Iceberg table may be located in different locations`, But when there is the corrupted table (metadata file is missing) then we won't able to perform `dropTableData` because we won't able to load the table.

So I think it is fine to use `unregister_table + delete table directory.`.

```java
        // Use the Iceberg routine for dropping the table data because the data files
        // of the Iceberg table may be located in different locations
        dropTableData(table.io(), table.operations().current());
```

findinpath replied:
> perhaps, dropping corrupted table could be done separately, as an unregister_table + delete table directory.

Do we have the certainty in such cases that the directory can be safely removed  or should this operation be left to the system administrator?
IMO we are speaking here about a corner case which should probably be handled in a best effort manner.

I'd advocate to simply unregister the table.

findepi commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestCorruptedIcebergTable.java:
```diff
@@ -56,63 +65,41 @@ public void testCorruptedTableLocation()
 
         // Assert queries fail cleanly
-        assertThatThrownBy(() -> query("TABLE " + tableName))
-                .hasMessageContaining("Failed to open input stream for file:");
```
If you end-up using `assertQueryFails` (as in the second commit), use it also in the first commit.

krvikash replied:
When I use `assertQueryFails` then `QueryAssertions` does not handle the exception whose cause is not `TrinoException` that is why I ended up using `assertThatThrownBy`. Please let me know if `QueryAssertions` is supposed to handle exceptions whose cause is not `TrinoException`.

Exception which I am getting when query fails is `io.trino.testing.QueryFailedException`.

```java
assertQueryFails("TABLE " + tableName, "Failed to open input stream for file:.*");
```

```
java.lang.IllegalArgumentException: Exception does not have TrinoException cause

	at io.trino.testing.QueryAssertions.lambda$getTrinoExceptionCause$4(QueryAssertions.java:513)
	at java.base/java.util.Optional.orElseThrow(Optional.java:403)
	at io.trino.testing.QueryAssertions.getTrinoExceptionCause(QueryAssertions.java:513)
	at io.trino.testing.QueryAssertions.lambda$assertQueryFails$3(QueryAssertions.java:443)
	at org.assertj.core.api.ThrowingConsumer.accept(ThrowingConsumer.java:30)
	at org.assertj.core.api.AbstractAssert.catchOptionalAssertionError(AbstractAssert.java:882)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:992)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)
	at org.assertj.core.api.AbstractAssert.satisfiesForProxy(AbstractAssert.java:873)
	at org.assertj.core.api.AbstractAssert.satisfies(AbstractAssert.java:862)
	at io.trino.testing.QueryAssertions.assertQueryFails(QueryAssertions.java:443)
	at io.trino.testing.AbstractTestQueryFramework.assertQueryFails(AbstractTestQueryFramework.java:436)
	at io.trino.plugin.iceberg.TestCorruptedIcebergTable.testCorruptedTableLocation(TestCorruptedIcebergTable.java:67)
```

krvikash replied:
Added `assertQueryFailsIncorrectly` in `TestCorruptedIcebergTable.java`

findepi commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/rest/TestIcebergTrinoRestCatalogConnectorSmokeTest.java:
```diff
@@ -173,7 +178,7 @@ public void testRegisterTableWithDifferentTableName()
     {
         assertThatThrownBy(super::testRegisterTableWithMetadataFile)
-                .hasMessageContaining("metadata location for register_table is not supported");
+                .hasMessageContaining("registerTable is not supported for Iceberg REST catalog");
```
Looks like a fix on its own. Separate commit?

findepi commented on testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergSparkCompatibility.java:
```diff
@@ -982,7 +982,7 @@ public void testCreateAndDropTableWithSameLocationFailsOnTrino(int specVersion)
 
         assertQueryFailure(() -> onTrino().executeQuery(format("SELECT * FROM %s", trinoTableName(tableSameLocation2))))
-                .hasMessageMatching(".*Failed to open input stream for file.*");
+                .hasMessageMatching(".*Metadata not found in metadata location for table (.*).*");
```
Does the exception still include the file on which we failed, along with the original exception?
let's test for that

findinpath commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/CorruptedIcebergTableHandle.java:
```diff
@@ -0,0 +1,36 @@
+    public CorruptedIcebergTableHandle
+    {
+        requireNonNull(schemaTableName, "schemaTableName is null");
+        requireNonNull(originalException, "originalException is null");
```
The name `originalException` in the context of a table handle is rather misleading.

Just another option to take into consideration: `originalException` -> `corruptionCause`

krvikash replied:
I was following delta-lake for the name https://github.com/trinodb/trino/blob/c23ec5e7d11edb471e8b26dd68dac0a067322b50/plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/CorruptedDeltaLakeTableHandle.java#L30

krvikash commented:
Thanks, @findepi | @findinpath for the reviews. Addressed comments.

krvikash commented:
Fixed CI failure

findinpath commented:
Tests are red

```
023-03-28T10:37:36.1282148Z 2023-03-28T04:37:35.895-0600	ERROR	http-worker-659	org.apache.iceberg.rest.RESTCatalogServlet	Error processing REST request
2023-03-28T10:37:36.1283043Z org.apache.iceberg.exceptions.RESTException: Unhandled error: ErrorResponse(code=404, type=NoSuchTableException, message=Table does not exist: tpch.test_drop_table_with_missing_snapshot_file_kljh1kx0xl)
2023-03-28T10:37:36.1283886Z org.apache.iceberg.exceptions.NoSuchTableException: Table does not exist: tpch.test_drop_table_with_missing_snapshot_file_kljh1kx0xl
2023-03-28T10:37:36.1284622Z 	at org.apache.iceberg.BaseMetastoreCatalog.loadTable(BaseMetastoreCatalog.java:50)
2023-03-28T10:37:36.1285554Z 	at org.apache.iceberg.rest.CatalogHandlers.loadTable(CatalogHandlers.java:240)
2023-03-28T10:37:36.1286219Z 	at org.apache.iceberg.rest.RESTCatalogAdapter.handleRequest(RESTCatalogAdapter.java:336)
2023-03-28T10:37:36.1286891Z 	at org.apache.iceberg.rest.RESTCatalogAdapter.execute(RESTCatalogAdapter.java:384)
2023-03-28T10:37:36.1287535Z 	at org.apache.iceberg.rest.RESTCatalogServlet.execute(RESTCatalogServlet.java:100)
2023-03-28T10:37:36.1288150Z 	at org.apache.iceberg.rest.RESTCatalogServlet.doGet(RESTCatalogServlet.java:66)
2023-03-28T10:37:36.1288707Z 	at javax.servlet.http.HttpServlet.service(HttpServlet.java:503)
2023-03-28T10:37:36.1289213Z 	at javax.servlet.http.HttpServlet.service(HttpServlet.java:590)
```

```
2023-03-28T10:44:47.9500066Z 2023-03-28T04:44:47.907-0600	ERROR	pool-3-thread-1	io.trino.testng.services.ProgressLoggingListener	[TEST FAILURE] io.trino.plugin.iceberg.TestIcebergMinioOrcConnectorSmokeTest.testDropTableWithMissingSnapshotFile; (took: 0.9 seconds)
2023-03-28T10:44:47.9501766Z io.trino.testing.QueryFailedException: Failed to get status for file: s3://test-iceberg-minio-smoke-test-vqhe3hnf6z/tpch_orc/test_drop_table_with_missing_snapshot_file_a05pfyc9tt-8c2da560b5424ec1bb46ec83f3cb44bf/metadata/snap-6691070755498700963-1-bd1b58ee-204d-4044-80bd-a02170234d30.avro
2023-03-28T10:44:47.9508936Z 	at io.trino.testing.AbstractTestingTrinoClient.execute(AbstractTestingTrinoClient.java:122)
2023-03-28T10:44:47.9509565Z 	at io.trino.testing.DistributedQueryRunner.executeWithQueryId(DistributedQueryRunner.java:494)
2023-03-28T10:44:47.9510306Z 	at io.trino.testing.QueryAssertions.assertDistributedUpdate(QueryAssertions.java:107)
2023-03-28T10:44:47.9510869Z 	at io.trino.testing.QueryAssertions.assertUpdate(QueryAssertions.java:63)
2023-03-28T10:44:47.9511441Z 	at io.trino.testing.AbstractTestQueryFramework.assertUpdate(AbstractTestQueryFramework.java:401)
2023-03-28T10:44:47.9512141Z 	at io.trino.testing.AbstractTestQueryFramework.assertUpdate(AbstractTestQueryFramework.java:396)
2023-03-28T10:44:47.9512961Z 	at io.trino.plugin.iceberg.BaseIcebergConnectorSmokeTest.testDropTableWithMissingSnapshotFile(BaseIcebergConnectorSmokeTest.java:536)
2023-03-28T10:44:47.9513708Z 	at java.base/jdk.int
```

krvikash commented:
Fixed CI failure.

krvikash commented:
Hi @ebyhr | @findepi, Could you please run the CI with secrets?

krvikash commented:
DeltaLake CI failure https://github.com/trinodb/trino/actions/runs/4553958406/jobs/8035260975 fixed by https://github.com/trinodb/trino/pull/16771

Rebased with the master.

krvikash commented:
Hi @yunrougong, FYI Alternatively, You can use the `unregister_table` procedure to drop the table from the metastores without deleting the data and delete data manually from the table location. 

https://trino.io/docs/current/connector/iceberg.html?highlight=unregister_table#unregister-table

krvikash commented:
Hi @ebyhr, could you please run the CI with secrets again?

ebyhr commented:
/test-with-secrets sha=5121a52fb421f3b71e9a31ba5cff0a75ea347f70

findinpath commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/AbstractIcebergTableOperations.java:
```diff
@@ -246,6 +249,23 @@ protected void refreshFromMetadataLocation(String newLocation)
+        if (failure.getCause() != null) {
+            Throwable cause = failure.getCause();
+            if (cause instanceof FileNotFoundException
+                    || cause instanceof TrinoS3FileSystem.UnrecoverableS3OperationException
```
`TrinoS3FileSystem.UnrecoverableS3OperationException` we shouldn't do such low-level checks here.

krvikash replied:
This low-level handling is purely based on what `TrinoS3FileSystem` throws when `newStream` is called.

`FileNotFoundException` or `UnrecoverableS3OperationException` are `IOException` which get caught and wrapped with `RuntimeException` in `TableMetadataParser#read`.

That is why I had to use `failure.getCause()`. If we do not abort on `UnrecoverableS3OperationException` exception, Then it will retry the operation 20 times which is costly.

krvikash replied:
Removed this code and handle `FileNotFoundException` in [/ForwardingSeekableInputStream.java](https://github.com/trinodb/trino/pull/16674/commits/cef5d099878d9007a5d4dba859ad28d8db138148#diff-d2efd26a616b31a9e87eab43321c46a74a89d6ff4ad21104fbc5023cba4ca56f)

However, It is still needed to handle `TrinoS3FileSystem.UnrecoverableS3OperationException` exception.

findinpath replied:
Can you pls create a follow-up issue for this kind of situation?

krvikash replied:
https://github.com/trinodb/trino/issues/16908

findinpath commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestCorruptedIcebergTable.java:
```diff
@@ -0,0 +1,134 @@
+/*
```
Add @findepi  as a co-author here.

findinpath commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/AbstractIcebergTableOperations.java:
```diff
@@ -246,6 +249,23 @@ protected void refreshFromMetadataLocation(String newLocation)
+        if (failure instanceof org.apache.iceberg.exceptions.NotFoundException) {
+            return true;
+        }
+        if (failure.getCause() != null) {
```
> With Glue catalog, the metadata file is read using the TrinoS3FileSystem....

`org.apache.iceberg.hadoop.HadoopInputFile#newStream` already throws `NotFoundException`


I don't follow the reasoning behind the low-level check of the cause of the exception.

krvikash replied:
As discussed offline, we will be checking for the `FileNotFoundException` exception here for `abortOn` instead of converting java's `FileNotFoundException` to Iceberg's `NotFoundException` in `ForwardingSeekableInputStream`.

For other abortable exceptions, I have created an issue https://github.com/trinodb/trino/issues/16908

yunrougong commented:
I want to put my case here with more context if it helps. see below. 

` File "/usr/local/lib/python3.7/site-packages/trino/client.py", line 440, in process
    raise self._process_error(response["error"], response.get("id"))
trino.exceptions.TrinoQueryError: TrinoQueryError(type=INTERNAL_ERROR, name=GENERIC_INTERNAL_ERROR, message="Failed to read file: HdfsInputFile{delegate=s3://test/data_lake/dw/iceberg/raw/reddit_test/metadata/00001-0f709f8c-e550-4642.metadata.json, identity=ConnectorIdentity{user='*****', groups=[], principal=*****, enabledSystemroles=[], extraCredentials=[]}}", query_id=20230330_013443_01387)`

The code throw this error because I manually deleted the data and metadata folder in staging ( dev)  S3, but I forgot to delete the corresponding registered AWS Glue Catalog Table schema.  After I cleaned up the corresponding glue catalog schema, I could be able to drop the tables via trino client.

krvikash commented:
Addressed comments and rebased with master.

krvikash commented:
> I want to put my case here with more context if it helps. see below.
> 
> ` File "/usr/local/lib/python3.7/site-packages/trino/client.py", line 440, in process raise self._process_error(response["error"], response.get("id")) trino.exceptions.TrinoQueryError: TrinoQueryError(type=INTERNAL_ERROR, name=GENERIC_INTERNAL_ERROR, message="Failed to read file: HdfsInputFile{delegate=s3://test/data_lake/dw/iceberg/raw/reddit_test/metadata/00001-0f709f8c-e550-4642.metadata.json, identity=ConnectorIdentity{user='*****', groups=[], principal=*****, enabledSystemroles=[], extraCredentials=[]}}", query_id=20230330_013443_01387)`
> 
> The code throw this error because I manually deleted the data and metadata folder in staging ( dev) S3, but I forgot to delete the corresponding registered AWS Glue Catalog Table schema. After I cleaned up the corresponding glue catalog schema, I could be able to drop the tables via trino client.

Thanks, @yunrougong for explaining the repro steps. I have added a test for the same.

https://github.com/trinodb/trino/pull/16674/files#diff-3f26b931a71ef29adbdba7ea025c513ecb76eeaa4578ad45cb3f05f254405762R589

krvikash commented:
Hi @ebyhr | @findinpath could you please take a look into this PR when you get time?

@ebyhr could you please run this PR with secrets?

ebyhr commented:
/test-with-secrets sha=e07f2aa4960a98256906e9404aae83e4ac4c3b81

findinpath commented:
@krvikash could you please rebase on `master` to make use of https://github.com/trinodb/trino/pull/16674 ?

findinpath commented on testing/trino-testing/src/main/java/io/trino/testing/AbstractTestQueryFramework.java:
```diff
@@ -442,6 +442,11 @@ protected void assertQueryFails(Session session, @Language("SQL") String sql, @L
         QueryAssertions.assertQueryFails(queryRunner, session, sql, expectedMessageRegExp);
     }
 
+    protected void assertQueryFailsIncorrectly(@Language("SQL") String sql, @Language("RegExp") String expectedMessageRegExp)
```
There are several methods here added and removed two commits later so that they can be used eventually within 
the test `TestCorruptedIcebergTable`:

https://github.com/trinodb/trino/pull/16674/commits/5d2c4eb0692595974a8a24c3d56bbaad35aacf1f

Add those methods locally to `TestCorruptedIcebergTable`.

ebyhr commented:
Regarding "Remove unused QueryAssertions#assertQueryFailsIncorrectly" & "Add QueryAssertions#assertQueryFailsIncorrectly", I don't think polluting commit history in testing modules only for Iceberg is a good idea. Can we add a helper method to `TestCorruptedIcebergTable` instead?

krvikash commented:
Addressed comments.

krvikash commented:
Rebased with master.

findinpath commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestCorruptedIcebergTable.java:
```diff
@@ -0,0 +1,139 @@
+                fail(format("Expected query to fail: %s [QueryId: %s]", sql, resultWithQueryId.getQueryId()));
+            }
+            else {
+                queryRunner.execute(session, sql);
```
This is dead code.

The `createQueryRunner()` method returns a `DistributedQueryRunner`.
```
IcebergQueryRunner.builder()
                .setMetastoreDirectory(dataDirectory)
                .build()
```

findinpath commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/hms/TrinoHiveCatalog.java:
```diff
@@ -333,12 +324,41 @@ public void dropTable(ConnectorSession session, SchemaTableName schemaTableName)
+            dropTableData(table.io(), metadata);
+        }
+        catch (RuntimeException e) {
+            // If the snapshot file is not found, an exception will be thrown by the dropTableData function.
```
Same question as in the case of TrinoGlueCatalog.

krvikash replied:
https://github.com/trinodb/trino/pull/16674#discussion_r1157285121

krvikash commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalog.java:
```diff
@@ -364,10 +365,29 @@ public void dropTable(ConnectorSession session, SchemaTableName schemaTableName)
         }
-        dropTableData(table.io(), table.operations().current());
+        try {
+            dropTableData(table.io(), table.operations().current());
```
I am expecting an exception not from `table.operations.current()` but from `dropTableData` method. 
`dropTableData` basically gets the snapshots from current metadata and iterates over it. with each snapshot files it loads all manifest files associated with it. If snapshot file is missing then it will throw `org.apache.iceberg.exceptions.NotFoundException` exception.

use case: When a snapshot file is missing then `dropTableData` will throw `org.apache.iceberg.exceptions.NotFoundException`. [`testDropTableWithMissingSnapshotFile`](https://github.com/trinodb/trino/pull/16674/files#diff-3f26b931a71ef29adbdba7ea025c513ecb76eeaa4578ad45cb3f05f254405762R520) covers this case.

findinpath replied:
Thank you for pointing this out.

findinpath commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:
```diff
@@ -1780,7 +1793,7 @@ public TableStatisticsMetadata getStatisticsCollectionMetadataForWrite(Connector
         }
 
-        IcebergTableHandle tableHandle = getTableHandle(session, tableMetadata.getTable(), Optional.empty(), Optional.empty());
+        ConnectorTableHandle tableHandle = getTableHandle(session, tableMetadata.getTable(), Optional.empty(), Optional.empty());
```
We can get a `CorruptedIcebergTableHandle` here.
Let's check whether we're dealing with a valid table handle before proceeding with the collection of statistics.

findinpath commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:
```diff
@@ -2685,10 +2698,10 @@ else if (strings.size() != 2) {
-            if (tableHandle == null) {
-                // Base table is gone
+            if (tableHandle == null || tableHandle instanceof CorruptedIcebergTableHandle) {
+                // Base table is gone or table is corrupted
```
A corrupted table can't be used anymore. Shouldn't we fail fast in such a situation?

findinpath commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/rest/TrinoRestCatalog.java:
```diff
@@ -228,6 +228,14 @@ public void dropTable(ConnectorSession session, SchemaTableName schemaTableName)
+    {
+        // Since it is currently not possible to obtain the table location, even if we drop the table from the metastore,
+        // it is still impossible to delete the table location.
+        throw new TrinoException(NOT_SUPPORTED, "dropCorruptedTable is not supported for Iceberg REST catalogs");
```
`"dropCorruptedTable is not supported for Iceberg REST catalogs"` -> "Cannot drop corrupted table %s from Iceberg REST catalog"

This is intended to be a more end-user friendly message.

krvikash commented:
Addressed comments.

krvikash commented:
rebased with master.

krvikash commented:
Fixed CI failure.

krvikash commented:
rebased with master and resolved conflicts

krvikash commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestCorruptedIcebergTable.java:
```diff
@@ -0,0 +1,135 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
```
Because The test is copied from [`TestDeltaLakeBasic#testCorruptedTableLocation`](https://github.com/trinodb/trino/blob/master/plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeBasic.java#L143)

findepi replied:
thanks for explaining. it's fine not to have me as a co-author

krvikash commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorSmokeTest.java:
```diff
@@ -484,6 +496,114 @@ public void testFileSortingWithLargerTable()
     }
 
+    @Test
+    public void testDropTableWithMissingMetadataFile()
```
Sorry, I could not understand the comment. Could you please explain a bit?

>the connector test.

Do you mean `BaseConnectorSmokeTest.java`?

ebyhr replied:
I guess it means `BaseConnectorTest`.

krvikash replied:
My purpose for adding the test cases in smoke test is because it covers all catalogs. whereas `BaseIcebergConnectorTest` only covers `HiveCatalog` and `GlueCatalog`.

> should generally have <= coverage

In this case, should the tests be duplicated in `BaseIcebergConnectorTest`?

krvikash commented:
Addressed comments.

findepi commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestCorruptedIcebergTable.java:
```diff
@@ -0,0 +1,133 @@
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+import static org.testng.Assert.fail;
+
+public class TestCorruptedIcebergTable
```
Do we want this test here?
or would BaseIcebergConnectorTest suffice?

krvikash replied:
Moved to `BaseIcebergConnectorTest`

krvikash commented:
Addressed comments.

findinpath commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergMinioOrcConnectorTest.java:
```diff
@@ -170,6 +171,12 @@ public void testDropAmbiguousRowFieldCaseSensitivity()
+    @Override
+    public void testCorruptedTableLocation()
+    {
+        throw new SkipException("Skipping test, This test override will be removed in next commit");
```
nit It's not clear from the exception why the skipping is being done - even if it is temporary.

findinpath commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorSmokeTest.java:
```diff
@@ -484,6 +496,114 @@ public void testFileSortingWithLargerTable()
+        String tableLocation = getTableLocation(tableName);
+
+        // Delete current metadata file
+        trinoFileSystem.deleteFile(metadataLocation);
```
I think that in such cases the columns of the table don't get taken into account while querying `information_schema.columns`

https://github.com/trinodb/trino/blob/0c99af47f6cee100d869fc83b72fa956fb4b743e/plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java#L642-L646

findinpath commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java:
```diff
@@ -6708,47 +6715,162 @@ public void testAlterTableWithUnsupportedProperties()
     }
 
+    @Test
+    public void testDropTableWithMissingMetadataFile()
```
we have these tests on both the Iceberg BCT as well as on Iceberg BCST

why?

findepi replied:
BCT should have no less coverage than BCST

we need BCST so that all Catalog impls are exercised.

findinpath commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java:
```diff
@@ -6708,47 +6715,162 @@ public void testAlterTableWithUnsupportedProperties()
+    }
+
+    @Test
+    public void testDropTableWithMissingSnapshotFile()
```
We should also make sure that we don't delete on `DROP` more than the content associated to the corrupted table.

I'm thinking that a test where two tables exist before the drop:

- one table is ok 
- one table is corrupt

When dropping the corrupt table, the OK table should still be present in the metastore and the amount of files in the storage (within the test schema) should decrease with only the number of files corresponding to the corrupted table.

findepi replied:
having the test as part of BCT / BCST provides that. there are other test tables (like nation) which don't get deleted.

findinpath reviewed (commented):
LGTM % comments

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalog.java:
```diff
@@ -364,10 +365,29 @@ public void dropTable(ConnectorSession session, SchemaTableName schemaTableName)
+            // If the snapshot file is not found, an exception will be thrown by the dropTableData function.
+            // So log the exception and continue with deleting the table location
+            LOG.warn(e, "Failed to delete table data referenced by metadata");
+        }
```
At this point we don't know what state the table data is, so leaving the table in metastore would be also problematic.
